pyquil/api/_devices.py

```diff
@@ -16,35 +16,66 @@
 from pyquil.device import Device
+from pyquil.api._errors import UnknownApiError
 
 
 def list_devices():
     """
-    Query the Forest 2.0 server for its knowledge of QPUs.
+    Query the Forest 2.0 server for a list of underlying QPU devices.
 
-    :return: A dictionary, keyed on device names. Each value is a dictionary of the form
-    {
-        "is_online":   a boolean indicating the availability of the device,
-        "is_retuning": a boolean indicating whether the device is busy retuning,
-        "specs":       a Specs object describing the entire device, serialized as a dictionary,
-        "isa":         an ISA object describing the entire device, serialized as a dictionary,
-        "noise_model": a NoiseModel object describing the entire device, serialized as a dictionary
-    }
+    NOTE: These can't directly be used to manufacture pyQuil Device objects, but this gives a list
+          of legal values that can be supplied to list_lattices to filter its (potentially very
+          noisy) output.
+
+    :return: A list of device names.
     """
+    # For the record, the dictionary stored in "devices" that we're getting back is keyed on device
+    # names and has this structure in its values:
+    #
+    # {
+    #   "is_online":   a boolean indicating the availability of the device,
+    #   "is_retuning": a boolean indicating whether the device is busy retuning,
+    #   "specs":       a Specs object describing the entire device, serialized as a dictionary,
+    #   "isa":         an ISA object describing the entire device, serialized as a dictionary,
+    #   "noise_model": a NoiseModel object describing the entire device, serialized as a dictionary
+    # }
+
     session = get_session()
     config = PyquilConfig()
 
-    return get_json(session, config.forest_url + "/devices")["devices"]
+    return list(get_json(session, config.forest_url + "/devices")["devices"].keys())
 
 
-def get_device(name: str):
+def list_lattices(device_name: str = None, num_qubits: int = None):
     """
-    Construct a Device object from a QCS-available device.
+    Query the Forest 2.0 server for its knowledge of lattices.  Optional filters by underlying
+    device name and lattice qubit count.
 
-    :param name: Name of the desired device.
-    :return: A Device object.
+    :return: A dictionary keyed on lattice names and valued in dictionaries of the form
+             {
+               "device_name": device_name,
+               "qubits": num_qubits
+             }
     """
-    raise NotImplementedError('QPU devices will be available at a later date, please use the QVM')
+    session = get_session()
+    config = PyquilConfig()
+
+    try:
+        response = get_json(session,
+                            config.forest_url + "/lattices",
+                            params={"device_name": device_name,
+                                    "num_qubits": num_qubits})
+
+        return response["lattices"]
+    except:
```

Comment:
"bare" excepts are usually bad. Can we either:

1. figure out the exception type of an unauthenticated user and only `except` those
2. `except Exception as e` and include the text of the base exception somewhere in the `ValueError`

Reply:
I'd prefer to do the latter, if you don't mind. Users trying to use this function without an endpoint configured, for instance, see a different exception than users who know the endpoint but don't have the credentials. The scope of this `try` is fairly narrow, but a lot of differently-named things could do wrong.